### PR TITLE
k6 backend

### PR DESCRIPTION
**Pull request by walterwanderley**
Add https://k6.io/ as load generator.

### CODE REVIEW

File: charts/kangal/README.md [hunk 1]
```diff
@@ -186,3 +188,11 @@ Deployment specific configurations:
 | `controller.env.GHZ_MASTER_CPU_REQUESTS`    | CPU requests    |                   |
 | `controller.env.GHZ_MASTER_MEMORY_LIMITS`   | Memory limits   |                   |
 | `controller.env.GHZ_MASTER_MEMORY_REQUESTS` | Memory requests |                   |
+
+### Kangal Controller (`k6` specific)
+| Parameter                                   | Description     | Default           |
+|---------------------------------------------|-----------------|-------------------|
+| `controller.env.K6_CPU_LIMITS`      | CPU limits      |                   |
```
Comment (s-radyuk) [inline charts/kangal/README.md:195]: I'd like to ask you to add these variables to docs/env-vars.md too.

File: docs/k6/README.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+More information regarding resource limits and requests can be found in the following page(s):
+
+- https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+- https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-resource-requests-and-limits
+
+## Writing tests
+It's recommended to read [official k6 documentation](https://k6.io/docs/).
+
+## Reporting
+k6 can write test statistics in multiple formats using the `K6_OUT` envvar. To persist the summary, put the code below into your testfile using the [handleSummary function](https://k6.io/docs/results-visualization/end-of-test-summary/#handlesummary-callback).
```
Comment (s-radyuk) [inline docs/k6/README.md:92]:
```suggestion
k6 can write test statistics in multiple formats using the `K6_OUT` environment variables. To persist the summary, put the code below into your testfile using the [handleSummary function](https://k6.io/docs/results-visualization/end-of-test-summary/#handlesummary-callback).
```

File: docs/k6/README.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+You can use https://github.com/benc-uk/k6-reporter to create a HTML Report.
+
+> Note: There aren't any concept of master/worker in k6. Metrics will not be automatically aggregated. To be able to aggregate your metrics and analyse them together, you’ll need to set K6_OUT env to send statics to another service (influxdb, prometheus, etc).
+
+## Logs
+
+There aren't any concept of master/worker in k6. All pods running k6 tests are workers.
+
+For the logs of the worker pod use the index number of the worker.
+Index numbers are `0`, `1`, etc, according to the number of workers you created.
```
Comment (s-radyuk) [inline docs/k6/README.md:127]:
```suggestion
Index numbers are `0`, `1`, etc., according to the number of workers you created.
```

File: pkg/backends/k6/resources.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+				},
+			},
+		})
+	}
+
+	args := make([]string, len(defaultArgs))
+	copy(args, defaultArgs)
+
+	if loadTest.Spec.Duration != 0 {
+		args = append(args, "--duration", loadTest.Spec.Duration.String())
```
Comment (s-radyuk) [inline pkg/backends/k6/resources.go:79]: Could you please add a note about using `duration` request parameter to docs/k6/README.md

**Review by s-radyuk — COMMENTED** [review]
Hello @walterwanderley 
Thank you very much for your contribution to Kangal! New backend looks good, I left some comments related to minor things. 
Also please rebase/merge master, the lint issues on opeapi.json were fixed there.

**Comment by s-radyuk** [comment]
I'd also would like to attract your attention to the following behaviour:
**How to reproduce:** Create a loadtest with several distributed pods and provide a duration.
```
curl --location --request POST 'http://0.0.0.0:8080/load-test' \
--header 'Content-Type: application/json' \
--form 'testFile=@"/script.js"' \
--form 'distributedPods="4"' \
--form 'type="K6"' \
--form 'duration="240s"'
```
script.js is an example file from [k6 documentation](https://k6.io/docs/getting-started/running-k6/#running-local-tests)

**Expected result:** all pods are running during the same duration.
**Actual result:** all pods started but only one pod is running according to the desired duration. Other pods immediately change status to Completed. 

If this is an expected behaviour I think this should be added to the documentation saying that it doesn't make sense to use more than one pod.

**Comment by walterwanderley** [comment]
> I'd also would like to attract your attention to the following behaviour: **How to reproduce:** Create a loadtest with several distributed pods and provide a duration.
> 
> ```
> curl --location --request POST 'http://0.0.0.0:8080/load-test' \
> --header 'Content-Type: application/json' \
> --form 'testFile=@"/script.js"' \
> --form 'distributedPods="4"' \
> --form 'type="K6"' \
> --form 'duration="240s"'
> ```
> 
> script.js is an example file from [k6 documentation](https://k6.io/docs/getting-started/running-k6/#running-local-tests)
> 
> **Expected result:** all pods are running during the same duration. **Actual result:** all pods started but only one pod is running according to the desired duration. Other pods immediately change status to Completed.
> 
> If this is an expected behaviour I think this should be added to the documentation saying that it doesn't make sense to use more than one pod.

Hello @s-radyuk 
Thanks for the review.

You need to configure `VUs >= DistributedPods` to put all pods to work. You can define VUs by exporting options object inside javascript code or by setting the `K6_VUS` environment variable.

I've added some documentation about this in the latest commit.

**Comment by lucasmdrs** [comment]
hey @walterwanderley :wave: 
Thank you for this awesome contribution!

Could you please include the new backend docs in the [mkdocs.yaml](https://github.com/hellofresh/kangal/blob/master/mkdocs.yaml) file?

**Comment by walterwanderley** [comment]
> hey @walterwanderley 👋 Thank you for this awesome contribution!
> 
> Could you please include the new backend docs in the [mkdocs.yaml](https://github.com/hellofresh/kangal/blob/master/mkdocs.yaml) file?

Hi @lucasmdrs, awesome project!

Ok! I've added k6 to mkdocs.

**Comment by lucasmdrs** [comment]
> Ok! I've added k6 to mkdocs.

Thanks! LGMT :heavy_check_mark:  

@s-radyuk do you have any other concerns ?

**Review by lucasmdrs — APPROVED** [review]

**Review by s-radyuk — APPROVED** [review]
LGTM 👍  Thank you for your contribution!

**Pull request merged**